### ISSUE TITLE
Feedback and questions

### ISSUE THREAD

**Issue by manuel-uberti** [minad/minicomp#1]
Hi,

first of all thanks for this package. I have been trying it out since yesterday and so far it's nice a experience indeed, since it perfectly fits my basic minibuffer completion needs.

A couple of questions:

- Do you plan to bind `C-p` and `C-n` to move around the candidates? As of now I can only do it with `<up>` and `<down>` (it could very well be my configuration, though I don't have this problem with `selectrum`)
- Do you plan to make the counter optional? I rarely need it. For instance, with `selectrum` I only use it with `consult-imenu` and `consult-ripgrep`

**Comment by minad** [comment]
> first of all thanks for this package. I have been trying it out since yesterday and so far it's nice a experience indeed, since it perfectly fits my basic minibuffer completion needs.

Thanks for the feedback. It is basically an alternative to icomplete without the cycling and the improved ivy/selectrum ui. But I don't plan to add more features (Selectrum quick keys, display actions etc).

> Do you plan to make the counter optional?

Yes, I have added this already.

> Do you plan to bind C-p and C-n to move around the candidates?

I added this too now. I missed this, since I am using ijkl navigation keys and these already worked because they get translated to arrows.

**Comment by manuel-uberti** [comment]
Fantastic, thank you for the quick feedback. Everything as expected, so we can close here. I'll keep you posted if I find anything else, but this already a great package.

**Comment by manuel-uberti** [comment]
@minad Without opening another ticket just for this: consider (M)ELPA for this, because it really is something I'd use regularly and I guess I will not be the only one. :)

**Comment by minad** [comment]
Yes, I am considering it. Maybe it makes sense to submit this to ELPA or even upstream as icomplete alternative. I am not sure yet. Maybe we find a better name, vcomplete, ...?

**Comment by manuel-uberti** [comment]
Ironically, `vcomplete` was the name of a macro I was using a while go when I was trying `selectrum` horizontal display but I still wanted some commands to output a vertical list of candidates.

Anyway, you are right, the name should suggest the vertical side of it. I can only think of `perpendicular` right now, but it sounds too high-brow already and doesn't add anything related to completion. Names are hard.

**Comment by jaor** [comment]
Hi @minad, i've also been trying this and it works really well for almost all my needs: having all this functionality in such a small package is really satisfying, and i also think it'd be nice to have minicomp in one elpa or another.  

The only thing that is not working for me at all are consult asyncs: is that because mincomp doesn't support dynamic tables?

tall-complete?

**Comment by minad** [comment]
> The only thing that is not working for me at all are consult asyncs: is that because mincomp doesn't support dynamic tables?

The package has full Consult and Embark support. `consult-ripgrep` etc works for me. What does not work for you? Maybe create a separate bug report with a more detailed description and please try emacs -Q.

**Comment by jaor** [comment]
On Sun, Apr 04 2021, Daniel Mendler wrote:

>     The only thing that is not working for me at all are consult asyncs: is that because mincomp doesn't
>     support dynamic tables?
>
> The package has full Consult and Embark support. consult-ripgrep etc
> works for me. What does not work for you?  Maybe create a separate bug
> report with a more detailed description and please try emacs -Q.

yes, i'll do, sorry about the vagueness (moreover, i'm using emacs 28).

i seem to need to set completion-styles to '(orderless) for things to
work at all (with the default, i get no completion UI at all), and when
i set that, the nifty # split operator is what doesn't work (it seems to
be used as part of the token to match, so i was getting empty results).

**Comment by minad** [comment]
> yes, i'll do, sorry about the vagueness (moreover, i'm using emacs 28).

Okay, I don't know if there is an issue because of Emacs 28. It is too much effort for me to test this. I will stay for a while on 27.1.

> i seem to need to set completion-styles to '(orderless) for things to
work at all (with the default, i get no completion UI at all), and when
i set that, the nifty # split operator is what doesn't work (it seems to
be used as part of the token to match, so i was getting empty results).

I just tested both `(setq completion-styles '(basic)` and `(setq completion-styles '(substring))`. Both styles work with `consult-ripgrep`. In the case of `basic` the string after the second `#` must match a prefix however, so I don't really consider this to be a usable completion style. Furthermore `basic` does not work with `consult-line` because of the disambiguation prefixes. For `consult-line` one needs a style which matches substrings in any case, something like substring, flex or orderless.

**Comment by jaor** [comment]
thanks for checking @minad.  i've played a bit with an emacs -q and
there's something in my configuration causing most of the trouble, i
think, because that way it mostly works (besides the caveats you
mention).

a piece of config malarkey i have is this:

      (defun jao-minicomp--orderless (old &rest args)
        (let ((completion-styles '(orderless)))
          (apply old args)))

      (advice-add 'minicomp--setup :around #'jao-minicomp--orderless)
      (advice-add 'minicomp--exhibit :around #'jao-minicomp--orderless)

what i want with that is to use orderless /only/ for minibuffer
completion (where i find it excellent), because i like company for
in-buffer completion but with the default completion styles rather than
orderless.

is there anything obviously wrong with the above?

**Comment by minad** [comment]
My recommendation is to use a minibuffer-setup-hook and there setq-local the desired completion style. The hook should run before minicomp--setup. Alternatively use a before advice on minicomp--setup. Setq-local is better than let binding, since let binding does not work well with recursive minibuffer sessions.

**Comment by jaor** [comment]
yes, a minibuffer-setup-hook is working well (i was unsure about
ordering issues, since that's a quite crowded hook in my system), and
has the nice property of actually being independent of minicomp.

also, after an update and recompilation of emacs 28, things are working
well everywhere for me with minicomp, including all my consult and
embark add-ons.  as already mentioned, i am very happy about minicomp
extremely high loc over functionality ratio.  definitely a package worth
publishing somewhere.

**Comment by jaor** [comment]
a quick update: setting completion-styles in the minibuffer hook turns out
to be very sensitive to ordering when it comes to async consult
commands.  i think it's because consult--split-setup also uses
setq-local on completion-styles to do its job, and if the hook changing
that value acts in between it breaks splitting.  i am trying the before
advice on minicomp--setup to see if it's less brittle.

**Comment by minad** [comment]
@jaor Yes, I know. Incidentally I just fixed this.

@manuel-uberti @jaor What about vertico (VERTical Interactive COmpletion)?

**Comment by minad** [comment]
@jaor But in any case you should ensure that your completion style setting should happen before consult async. At least I removed the order dependence from mimicomp/vertico.

**Comment by jaor** [comment]
> @jaor Yes, I know. Incidentally I just fixed this.

excellent

> @manuel-uberti @jaor What about vertico (VERTical Interactive COmpletion)?

i like it! if you'll forgive the awful pun: it's so spiffy, it almost
gives you vertigo. quite apt.

**Comment by minad** [comment]
Yes, the pun is the fun of it :D

**Comment by manuel-uberti** [comment]
Nice name indeed!

**Comment by minad** [comment]
Submitted here https://lists.gnu.org/archive/html/emacs-devel/2021-04/msg00177.html, in case someone wants to voice some support ;)

**Comment by manuel-uberti** [comment]
Just did. Once again, thanks a lot for this package (and all the rest of your work of course!).

**Comment by minad** [comment]
Thanks. I saw that you used Embark live completions for a while. What made you stop using that? For me the main advantage of vertical completion systems like Selectrum/Vertico is that it allows a very fast selection and navigation, but maybe this is only perceived as such. For more complicated tasks I can always `embark-export` or `embark-collect-snapshot`. Selectrum should be even faster than Vertico since it optimizes static completion tables at the cost of a bit more complexity.

**Comment by manuel-uberti** [comment]
Coming from Helm, I want the completion system to offer the same experience, meaning I want the same user interface and user experience for all the completion-driven commands I regularly use, such as searches with `ripgrep`, browsing projects with `project.el`, etc. 

That's why I like the Consult+Embark+Orderless approach on top of Selectrum/vertico: I have the same experience I had with Helm, but a with simpler and easier to understand code base, at least to one that doesn't master ELisp.

With regard to Embark live completion, I prefer Selectrum and vertico because they *feel* snappier, but as you say maybe is just a weird perception.

**Comment by minad** [comment]
Thanks for your answer. My experience is similar. I like if packages push it as far as possible in terms of integration/compatibility with existing functionality. Enhancing components feels better than replacing components and is very well possible given the flexibility of Elisp. And this helps with understanding.

Stefan Monnier already added the package. This seems to be a quick process, maybe I will move more packages to ELPA.

**Comment by jaor** [comment]
On Mon, Apr 05 2021, Manuel Uberti wrote:

> That's why I like the Consult+Embark+Orderless approach on top of
> Selectrum/vertico: I have the same experience I had with Helm, but a
> with simpler and easier to understand code base, at least to one that
> doesn't master ELisp.
>
> With regard to Embark live completion, I prefer Selectrum and vertico
> because they feel snappier, but as you say maybe is just a weird
> perception.

FWIW, Manuel's experience and comments pretty much coincide with mine
(except that i came from the ivy camp).  I also tried embark completion,
because i liked the idea of depending on one less package and having a
system that, as a consequence, is more undertandable and hackable, but i
found it too slow in some situations.  I think embark's way works well
if one is happy with only seeing a list of completions on demand and not
too often (that's how Omar Antoline told me he uses it), but i'm too
fond of interfaces like selectrum's and now vertico's.  OTOH, the things
in selectrum that vertico lacks are precisely those i don't care about,
so vertico, with its understandable 500 loc, fits really well in my
emacs OS.

**Comment by minad** [comment]
Thanks for responding too, @jaor. I agree about Embark completion - Omar is happy with not seeing the completions directly and in this case Embark works well. Regarding Selectrum, I don't think it is bad that it provides options (quick keys and display actions), since some people like that. My main criticism is that it rolls its own filtering/sorting pipeline in order to achieve good performance. Unfortunately this requires quite a bit of work in order to achieve full compliance with all the Emacs completion tables, in particular dynamic completion tables (https://github.com/raxod502/selectrum/issues/481). This is certainly a detail and nothing that the end-user really cares about. Vertico is basically an experiment if it is good enough to just use the default infrastructure and combine it with optimized sorting, without adding any special cases. Actually there is a tiny bit of special casing in the sorting to handle files, which I am not super happy with, but I don't think there is another way. It could be that it turns out that Vertico's simple architecture breaks down for some edge cases and then Selectrum's architecture is in fact better since it controls more of the pipeline. But then, Icomplete uses an architecture very similar to Vertico, so maybe Vertico will just work well enough?

**Comment by jaor** [comment]
thanks for the rationale, @minad.

> It could be that it turns out that Vertico's simple architecture
> breaks down for some edge cases and then Selectrum's architecture is
> in fact better since it controls more of the pipeline. But then,
> Icomplete uses an architecture very similar to Vertico, so maybe
> Vertico will just work well enough?

i'm certainly hoping it will. a further benefit of its architecture is
that, should it break down, it might pinpoint needed improvements in
emacs's completion machinery itself, specially now that vertico is
included in gnu elpa.

**Comment by minad** [comment]
@ssnnoo The crucial difference is that Vertico recomputes the candidates every time completion since otherwise we cannot support dynamic completion tables properly. Selectrum basically has to do the same for dynamic tables, see https://github.com/raxod502/selectrum/pull/532. But for now Selectrum made the assumption that most tables are static and it may break for some dynamic tables.

Since Selectrum optimizes static tables it does better regarding candidate preprocessing. It does not have to preprocess and sort the candidates every time. In Vertico I disable sorting for more than 20k candidates, see `vertico-sort-threshold`. Vertico could also add some additional caching for static tables, but for now I have not found that to be necessary since I am fine with having the threshold.

Maybe @clemera can tell you more?

**Comment by clemera** [comment]
Selectrum did not support completion-styles at the beginning, as far as I can tell @raxod502 designed the filtering/sorting pipeline mainly with prescient in mind. The design is better than the built-in design where completion-styles do both filter and highlighting of all candidates, highlighting only the displayed candidates gives a great performance boost. I like that Selectrums design makes it easy to adjust filtering/sorting but there is also the downside is that these adjustments are Selectrum specific.

**Comment by minad** [comment]
I agree. If the system would be implemented from scratch one would probably go with something more into the direction of Selectrum. But given that the existing infrastructure makes so many assumptions and basically forces you into a design, it seems very hard to improve on that.

**Comment by bdarcus** [comment]
Wow; didn't expect this!

Can I ask a question here @minad? You won't be surprised by this, but what are your plans on the crm UI?

**Comment by minad** [comment]
> Can I ask a question here @minad? You won't be surprised by this, but what are your plans on the crm UI?

Yes, I know you want an improvement regarding crm. Right now I don't have such plans. Vertico aims to minimalistic and uses the built-in completion tables as is (crm is just another completion table). I experimented a bit in the crm* branches, but nothing conclusive.

**Comment by bdarcus** [comment]
OK, understood.

FWIW, I just added a vertico screenshot to my [README](https://github.com/bdarcus/bibtex-actions#what-is-it), just to demonstrate it works with solutions other than selectrum.

**Comment by minad** [comment]
Looks pretty nice your readme :)

Regarding CRM if you are interested you can also play around and experiment with this. There are few low-cost improvements one can do:

https://github.com/minad/vertico/blob/d5556b8bc25755d097cfa5feb4acb9da08dd7af1/README.org#enhancing-the-ui-of-completing-read-multiple

Then there is another idea:

You can write a function which you register as an `after-change-function` hook in the minibuffer. This function splits the already selected candidates and replaces the candidates with `(propertize candidate 'display "short-candidate-version")`. Then the default UI will work well with your bibtex-completions package despite having long candidates by default.

~~~ elisp
(let ((candidates '(("long name1" . "short1")
                    ("long name2" . "short2")
                    ("long name3" . "short3")
                    ("long name4" . "short4"))))
  (minibuffer-with-setup-hook
      (lambda ()
        (add-hook 'after-change-functions
                  (lambda (&rest _)
                    (save-excursion
                      (save-match-data
                        (let ((pos (minibuffer-prompt-end)))
                          (goto-char pos)
                          (while (< pos (point-max))
                            (when (re-search-forward (concat crm-separator "\\|\\'") nil 'noerror)
                              (when-let (short (cdr (assoc (string-trim (buffer-substring-no-properties
                                                                         pos (match-beginning 0)))
                                                           candidates)))
                                (put-text-property pos (match-beginning 0) 'display short))
                              (setq pos (point))))))))
                  nil 'local))
  (completing-read-multiple "CRM: " candidates nil t)))
~~~

**Comment by bdarcus** [comment]
OK, thanks.

Copying-and-pasting the code in the README doesn't seem to work, though. I end up with this, and if I type in some characters nothing returns:

![Screenshot from 2021-04-11 17-59-44](https://user-images.githubusercontent.com/1134/114322753-ea146b00-9aef-11eb-8c0e-7474dadb597f.png)

**Comment by bdarcus** [comment]
> Looks pretty nice your readme :)

Nice themes go a long way; `doom-one` and `modus-operandi` (for vertico).

**Comment by bdarcus** [comment]
@minad 

> You can write a function which you register as an `after-change-function` hook in the minibuffer. This function splits the already selected candidates and replaces the candidates with `(propertize candidate 'display "short-candidate-version")`.

I'm just getting back to trying to implement this idea, which does indeed seem to be a good one.

How would I adapt this to my case, where I have multiple commands using CRM? I'm confused by this:

``` elisp
  (minibuffer-with-setup-hook
...
  (completing-read-multiple "CRM: " candidates nil t)))
```

**Comment by minad** [comment]
> How would I adapt this to my case, where I have multiple commands using CRM? I'm confused by this:

You have to implement this separately per command.

**Comment by bdarcus** [comment]
> You have to implement this separately per command. Each `completing-read-multiple` call should be wrapped by `minibuffer-with-setup-hook` which sets up the after change hook.

As in, something like this?

``` elisp
(defun bibtex-actions-insert-key (keys)
  "Insert BibTeX KEYS.
With prefix, rebuild the cache before offering candidates."
  (minibuffer-with-setup-hook
    ...
    (interactive (list (bibtex-actions-read :rebuild-cache current-prefix-arg))))
  (bibtex-completion-insert-key keys))
```

**Comment by minad** [comment]
No, in the `bibtex-actions-read` function *directly* around the `completing-read-multiple` call.
